core/field_textinput.js

```diff
@@ -371,12 +366,30 @@ Blockly.FieldTextInput.prototype.widgetDispose_ = function() {
   var thisField = this;
   return function() {
     var htmlInput = Blockly.FieldTextInput.htmlInput_;
-    // Save the edit (if it validates).
-    thisField.maybeSaveEdit_();
 
+    // Finalize value.
+    thisField.isBeingEdited_ = false;
+    // No need to call setValue because if the widget is being closed the
+    // latest input text has already been validated.
+    if (thisField.value_ != thisField.text_) {
+      // At the end of an edit the text should be the same as the value. It
+      // may not be if the input text is different than the validated text.
+      // We should fix that.
+      thisField.text_ = String(thisField.value_);
+      thisField.isTextValid_ = true;
+      thisField.forceRerender();
+    }
+    // Otherwise don't rerender.
+
+    // Call onFinishEditing
+    // TODO: Get rid of this or make it less of a hack.
```

Comment:
FWIW the case where this is really important is when editing the name of a parameter in a function.  That has code to create a variable with the given name at every keypress, and then it deletes intermediate variables when the user is done editing.